File: composeApp/src/iosMain/kotlin/org/example/kmpsamples/presentation/permissions/systemPermissionHandlers/SystemPermissionHandlerLocator.kt
package org.example.kmpsamples.presentation.permissions.systemPermissionHandlers

import org.example.kmpsamples.permissions.systemPermissionHandlers.SystemPermissionHandler
import org.example.kmpsamples.presentation.permissions.PlatformPermissionMapper
import org.example.kmpsamples.presentation.permissions.dtos.PlatformPermissionType

interface SystemPermissionHandlerLocatorInterface {
    fun getPermissionHandler(
        permissionType: PlatformPermissionType
    ): SystemPermissionHandler
}

class SystemPermissionHandlerLocator : SystemPermissionHandlerLocatorInterface {
    private val systemPermissionHandlers =
        setOf(
            ContactsHandler(),
            RecordAudioHandler(),
            CalendarLegacyHandler(),
            CalendarWriteAccessIos17Handler(),
            CalendarFullAccessIos17Handler()
        )

    override fun getPermissionHandler(
        platformPermissionType: PlatformPermissionType
    ): SystemPermissionHandler {
        val singlePermissions =
            PlatformPermissionMapper.toPlatformSinglePermissions(listOf(platformPermissionType))
        return systemPermissionHandlers.find { handler ->
            singlePermissions.any { handler.canHandle(it) }
        }
            ?: error("No system permission handler can handle permission type: $platformPermissionType")
    }
}

File: iosApp/iosApp/Permissions/SystemPermissionHandlers/SystemPermissionHandlerLocator.swift
import ComposeApp

protocol SystemPermissionHandlerLocatorInterface {
    func getPermissionHandler(permissionType: PlatformPermissionType) -> SystemPermissionHandler
}

class SystemPermissionHandlerLocator: SystemPermissionHandlerLocatorInterface {
    private let systemPermissionHandlers: [SystemPermissionHandler] = [
        ContactsHandler(),
        RecordAudioHandler(),
        CalendarLegacyHandler()] +
    {
        if #available(iOS 17.0, *) {
            return [
                CalendarWriteAccessIos17Handler(),
                CalendarFullAccessIos17Handler()]
        } else {return []}
    }()
    
    func getPermissionHandler(permissionType: PlatformPermissionType) -> SystemPermissionHandler {
        let singlePermissions = PlatformPermissionMapper.toPlatformSinglePermissions([permissionType])
        
        if let handler = systemPermissionHandlers.first(where: { handler in
            singlePermissions.contains(where: { handler.canHandle(simplePlatformPermission: $0) })
        }) {
            return handler
        } else {
            fatalError("No system permission handler can handle permission type: \(permissionType)")
        }
    }
}
